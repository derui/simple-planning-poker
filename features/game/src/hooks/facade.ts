import { createFacade } from "@spp/shared-hook-facade";
import { UseCreateGame, UseListGame, UsePrepareGame } from "../atoms/game.js";

export type Hooks = {
  useCreateGame: UseCreateGame;
  usePrepareGame: UsePrepareGame;
  useListGame: UseListGame;
};

const { hooks, ImplementationProvider } = createFacade<Hooks>();

export { hooks, ImplementationProvider };
