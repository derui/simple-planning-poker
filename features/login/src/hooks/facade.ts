import { createFacade } from "@spp/shared-hook-facade";
import { UseAuth, UseLogin } from "../atoms/atom.js";

export type Hooks = {
  useAuth: UseAuth;
  useLogin: UseLogin;
};

const facade: ReturnType<typeof createFacade<Hooks>> = createFacade<Hooks>();

export const hooks: Readonly<Hooks> = facade.hooks;
export const ImplementationProvider: typeof facade.ImplementationProvider = facade.ImplementationProvider;
