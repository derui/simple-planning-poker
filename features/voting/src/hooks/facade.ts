import { createFacade } from "@spp/shared-hook-facade";
import { UseJoin, UsePollingPlace, UseRevealed, UseVoting, UseVotingStatus } from "../atoms/voting.js";

export type Hooks = {
  useJoin: UseJoin;
  useVoting: UseVoting;
  useRevealed: UseRevealed;
  useVotingStatus: UseVotingStatus;
  usePollingPlace: UsePollingPlace;
};

const facade: ReturnType<typeof createFacade<Hooks>> = createFacade<Hooks>();

export const hooks: Readonly<Hooks> = facade.hooks;
export const ImplementationProvider: typeof facade.ImplementationProvider = facade.ImplementationProvider;
