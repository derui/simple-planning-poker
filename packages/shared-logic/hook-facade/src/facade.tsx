import { createContext, useContext } from "react";

export type ImplementationProvider<T> = React.ComponentType<React.PropsWithChildren<{ implementation: T }>>;

/**
 * Create facade with type `Hooks`
 */
export const createFacade = function createFacade<Hooks extends Record<string, () => unknown>>(): {
  hooks: Readonly<Hooks>;
  ImplementationProvider: ImplementationProvider<Hooks>;
} {
  const providerNotFound: unique symbol = Symbol();

  const Context = createContext<Hooks | typeof providerNotFound>(providerNotFound);

  const implementationHandler: ProxyHandler<Hooks> = {
    get(_target, prop, _receiver) {
      // eslint-disable-next-line
      const obj = useContext(Context);

      if (obj === providerNotFound) {
        throw new Error("Provider not found");
      }

      if (typeof prop == "symbol") {
        throw new Error("Can not use symbol as hook key");
      }

      return obj[prop];
    },
    has() {
      return false;
    },
    ownKeys() {
      return [];
    },
    getOwnPropertyDescriptor() {
      return undefined;
    },
    getPrototypeOf() {
      return null;
    },
    preventExtensions() {
      return true;
    },
    isExtensible() {
      return false;
    },
    set() {
      return false;
    },
    deleteProperty() {
      return false;
    },
  };

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const proxy = new Proxy<Hooks>({} as Hooks, implementationHandler);

  const ImplementationProvider: ImplementationProvider<Hooks> = ({ implementation, children }) => {
    const consumed = useContext(Context);

    if (consumed !== providerNotFound) {
      throw new Error("Provider already defined");
    }

    return <Context.Provider value={implementation}>{children}</Context.Provider>;
  };

  return { hooks: proxy, ImplementationProvider };
};
