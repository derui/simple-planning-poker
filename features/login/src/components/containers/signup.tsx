import * as Url from "@spp/shared-app-url";
import { useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { hooks } from "../../hooks/facade.js";
import { SignInLayout } from "./signin.layout.js";

// eslint-disable-next-line func-style
export function SignUp() {
  const login = hooks.useLogin();
  const navigate = useNavigate();

  useEffect(() => {
    if (login.status == "logined") {
      navigate(Url.gameIndexPage());
    }
  }, [login.status]);

  return <SignInLayout title="Sign Up" onSubmit={login.signUp} loading={login.status == "doing"} />;
}
