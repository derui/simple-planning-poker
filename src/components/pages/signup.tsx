import classNames from "classnames";
import { useEffect } from "react";
import { useNavigate } from "react-router";
import { useAppDispatch, useAppSelector } from "../hooks";
import { SignIn } from "../presentations/signin";
import { Overlay } from "../presentations/overlay";
import { Loader } from "../presentations/loader";
import { signUp } from "@/status/actions/signin";
import { selectAuthenticated, selectAuthenticating } from "@/status/selectors/auth";

const styles = {
  overlayDialog: classNames(
    "flex",
    "flex-row",
    "p-4",
    "h-16",
    "border",
    "border-secondary1-400",
    "rounded",
    "bg-white",
    "items-center"
  ),
  dialogText: classNames("ml-3"),
};

// eslint-disable-next-line func-style
export function SignUpPage() {
  const authenticating = useAppSelector(selectAuthenticating());
  const authenticated = useAppSelector(selectAuthenticated());
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  const handleSubmit = ({ email, password }: { email: string; password: string }) => {
    dispatch(signUp({ email, password }));
  };

  useEffect(() => {
    if (authenticated) {
      navigate("/game", { replace: true });
    }
  }, [authenticated]);

  return (
    <div>
      <Overlay show={authenticating} testid={"overlay"}>
        <div className={styles.overlayDialog}>
          <Loader size="m" shown={true} testid={"loader"} />
          <span className={styles.dialogText}>Authenticating...</span>
        </div>
      </Overlay>
      <SignIn title="Sign Up" onSubmit={handleSubmit} authenticating={authenticating} testid="signin"></SignIn>
    </div>
  );
}

export default SignUpPage;
