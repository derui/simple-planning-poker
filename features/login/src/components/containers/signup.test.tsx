import { cleanup, render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { createStore, Provider } from "jotai";
import { MemoryRouter } from "react-router-dom";
import sinon from "sinon";
import { afterEach, expect, test } from "vitest";
import { Hooks, ImplementationProvider } from "../../hooks/facade.js";
import { SignUp } from "./signup.js";

afterEach(cleanup);

test("render page", () => {
  const store = createStore();

  const mock: Hooks = {
    useLogin() {
      return {
        signIn: sinon.fake(),
        signUp: sinon.fake(),
        status: "notLogined",
        loginError: undefined,
      };
    },
    useAuth: sinon.fake(),
  };

  render(
    <ImplementationProvider implementation={mock}>
      <Provider store={store}>
        <MemoryRouter>
          <SignUp />
        </MemoryRouter>
      </Provider>
    </ImplementationProvider>
  );

  expect(screen.getByRole("main")).not.toBeNull();
  expect(screen.queryByText(/Sign Up/)).not.toBeNull();
});

test("call hook after submit", async () => {
  // Arrange
  const signUpFake = sinon.fake();
  const mock: Hooks = {
    useLogin() {
      return {
        signIn: sinon.fake(),
        signUp: signUpFake,
        status: "notLogined",
        loginError: undefined,
      };
    },
    useAuth: sinon.fake(),
  };

  // Act
  render(
    <ImplementationProvider implementation={mock}>
      <MemoryRouter>
        <SignUp />
      </MemoryRouter>
    </ImplementationProvider>
  );

  await userEvent.type(screen.getByPlaceholderText("e.g. yourname@yourdomain.com"), "email");
  await userEvent.type(screen.getByPlaceholderText("Password"), "password");
  await userEvent.click(screen.getByText("Submit"));

  // Assert
  expect(signUpFake.lastCall.args).toEqual(["email", "password"]);
});
