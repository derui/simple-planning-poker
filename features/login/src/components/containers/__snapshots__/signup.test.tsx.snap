// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`call hook after submit 1`] = `
<div>
  <main
    class="signin_root__dhmk8q0"
  >
    <div
      class="style_hidden__gg6aol1 style_root__gg6aol0 style_transition_all__13d8cin41"
      data-show="false"
      role="dialog"
    >
      <div
        class="signin_overlayDialog__dhmk8q1"
      >
        <span
          class="style_loader_m__m7nytx3 style_base__m7nytx0"
          data-shown="true"
          role="alert"
        >
          <span
            class="style_spinLoader_m__m7nytx7 style_spinLoaderBase__m7nytx5 style_animation_spin__13d8cin3z"
          >
            <span
              class="loader2_medium_gray__1nawc259 style_iconMedium__11pykq32 style_iconBase__11pykq30"
            />
          </span>
        </span>
        <span
          class="signin_dialogText__dhmk8q2"
        >
          Authenticating...
        </span>
      </div>
    </div>
    <div
      class="style_root__18vd3zh0"
      role="dialog"
    >
      <div
        class="style_header__18vd3zh1"
      >
        Sign Up
      </div>
      <div
        class="style_main__18vd3zh2"
        role="article"
      >
        <div
          class="signin_dialogContent__dhmk8q3"
        >
          <form
            class="login-form_root__11r7lvx0"
          >
            <div
              class="login-form_inputContainer__11r7lvx2"
            >
              <label
                class="login-form_inputLabel__11r7lvx3"
              >
                Email
              </label>
              <input
                class="login-form_input__11r7lvx4 style_transition_all__13d8cin41"
                name="email"
                placeholder="e.g. yourname@yourdomain.com"
                tabindex="0"
                value="email"
              />
            </div>
            <div
              class="login-form_inputContainer__11r7lvx2"
            >
              <label
                class="login-form_inputLabel__11r7lvx3"
              >
                Password
              </label>
              <input
                class="login-form_input__11r7lvx4 style_transition_all__13d8cin41"
                minlength="6"
                name="password"
                placeholder="Password"
                tabindex="0"
                type="password"
                value="password"
              />
            </div>
            <div
              class="login-form_submitContainer__11r7lvx5"
            >
              <button
                class="style_enabledButtonVariants_gray__nv52ero style_enabledBase__nv52er3 style_base__nv52er0 style_transition_all__13d8cin41"
                type="button"
              >
                <span
                  class="style_hidden__m7nytx1"
                  data-shown="false"
                  role="alert"
                >
                  <span
                    class="style_spinLoader_s__m7nytx6 style_spinLoaderBase__m7nytx5 style_animation_spin__13d8cin3z"
                  >
                    <span
                      class="loader2_small_gray__1nawc250 style_iconSmall__11pykq31 style_iconBase__11pykq30"
                    />
                  </span>
                </span>
                Back
              </button>
              <button
                class="style_enabledButtonVariants_emerald__nv52err style_enabledBase__nv52er3 style_base__nv52er0 style_transition_all__13d8cin41"
                type="submit"
              >
                <span
                  class="style_hidden__m7nytx1"
                  data-shown="false"
                  role="alert"
                >
                  <span
                    class="style_spinLoader_s__m7nytx6 style_spinLoaderBase__m7nytx5 style_animation_spin__13d8cin3z"
                  >
                    <span
                      class="loader2_small_gray__1nawc250 style_iconSmall__11pykq31 style_iconBase__11pykq30"
                    />
                  </span>
                </span>
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
</div>
`;
